[TAS-434] 잔소리 대상 두윗 조회 API

S-J-Kim opened the pull request:
## PR 체크리스트

Merge 전에 아래 체크리스트가 모두 체크되었는지 확인해주세요

- [x] 마지막 commit전에 빌드하여 코드 스타일 점검하였나요?
- [ ] 모든 Test 코드를 실행하여 PASS했나요?
- [ ] Swagger 명세를 작성하였나요?
- [ ] 최소 1명의 리뷰와 Approve를 받았나요?

## PR 타입

PR의 타입을 체크해주세요

- [ ] 버그 픽스 - 이슈 링크 :
- [x] 신규 기능 개발
- [ ] 기능 수정
- [ ] 테스트 코드 추가
- [ ] 코드 스타일 업데이트 (formatting, 변수명 수정 등)
- [ ] 리팩토링 (기능 수정은 없고 코드 재사용성을 위한 메서드 분리 등)
- [ ] 설정 파일 수정
- [ ] 기타

## 백로그 및 이슈 링크

- 링크 :

## 변경된 사항

### ~~Ranking 도메인 수정~~

- ~~레이지 두윗러 목록도 저정할 수 있도록 수정~~
- ~~`year`, `month`, `week`: NOT NULL -> DEFAULT NULL 로 수정~~
- ~~`member_id` -> `entity_id` 로 수정~~
-> fallback은 db 로직을 통해 직접 수행하는 것으로 결정하여 Ranking은 수정하지 않음

### 레이지 두윗러 & 잔소리 대상 두윗 캐시 적재 배치 개발
- Domain service와 QueryDSL을 사용하여 개발함. #74 의 공통 모듈 리팩토링 구조 참고

### RedisOperator 개발
현재의 CacheHelper가 Redis를 캐싱과 data store의 용도를 모두 고려하여 구현되어있어 사용 시 혼란이 발생할 수 있다고 판단함. 순수 캐싱의 경우 CacheManager를 활용하고, data store가 필요한 Ops는 이 모듈을 사용하는 것을 제안하기 위해 구현.

- 순수 캐싱은 CacheHelper보다, 타입을 string으로 제한하고 애너테이션 기반으로 사용하는 것이 좋을 것 같음. (Spring Cache의 사상을 최대한 수용)
- Data store는 Spring Cache 사상 범위 밖에 있으므로 아예 RedisTemplate 기반의 Ops 헬퍼 유틸을 사용하는 것이 바람직하다는 의견

### CachePoliySpec 인터페이스 추가
- CachePolicy가 테스트 가능하도록 인터페이스 구현의 형태로 리팩토링함. (기존 코드 영향 없음)

## 기타 전달 사항
레이지 두윗러"만"을 위한 id 저장보다, 아예 모든 "잔소리 대상 두윗" 에 id를 저장하는 방식으로 생각을 바꿔보았습니다.
어차피 레이지 두윗러는 잔소리 대상 두윗의 부분집합이기 때문에, 모든 잔소리 대상 두윗은 레이지 두윗러의 잠재적 후보로 볼 수 있습니다.
초기 논의 시 top 15를 추출해 Redis에 적재하는 이유도, top 5 중 중간에 완료/삭제 되면 차순위를 top5에 포함시키기 위함인데, 이를 확장해서 아예 모든 두윗의 id를 인덱싱 해놓는다면 레이지 두윗러 목록이 5개 미만으로 내려갈 확률이 대폭 감소할 것 같습니다.

추가로 어차피 잔소리 대상 두윗을 모두 조회하기 위해선 id 전체 목록을 알고있거나, SCAN 연산을 해야합니다.
SCAN 연산으로 전체 id 목록 획득 후 -> HGET 으로 하는 것 보다 애초에 모든 id 목록을 미리 가지고 있어 바로 조회하는 것이 더 유리하다는 생각입니다.
 
## (참고) 배치 Sequence diagram
`LazyDowithSelector` 는 도메인 서비스

```mermaid
sequenceDiagram
    autonumber

    participant Tasklet
    participant LazyDowithSelector
    participant Redis
    participant DB

    %% 1) Dowith 조회
    Tasklet->>DB: 모든 잔소리 대상 두윗 조회
    DB-->>Tasklet: DowithList
    
    %% 2) 레이지 두윗러 추출
    Tasklet->>LazyDowithSelector: 레이지 두윗러 목록 추출
    LazyDowithSelector->>Tasklet: lazy dowith list

    %% 3) Dowith 개별 Redis 적재
    loop Store each dowith
        Tasklet->>Redis: HSET feed::dowithTask::{id} {...fields}
    end

		%% 4) 레이지 두윗 id는 원자적으로 갱신 (rename)
    Tasklet->>Redis: RPUSH feed::dowithTask::lazy::temp_{timestamp} [id1,id2,...]
    Tasklet->>Redis: RENAME feed::dowithTask::lazy::temp_{timestamp} feed::dowithTask::lazy

    Tasklet-->>Tasklet: Batch Completed

```

S-J-Kim committed:
feat: Ranking 테이블 구조 수정; 레이지 두윗러 목록도 적재 가능하도록 수정

S-J-Kim committed:
feat: 잔소리 대상 두윗 & 레이지 두윗러 캐시 적재 배치 (WIP)

S-J-Kim commented on src/main/java/com/LetMeDoWith/LetMeDoWith/batch/job/task/StoreFeedbackAvailableDowithTasksConfig.java:
```diff
@@ -0,0 +1,43 @@
+        return new StepBuilder(STEP_NAME, jobRepository)
+                .tasklet(tasklet, platformTransactionManager)
+                .build();
+    }
```
`Job`의 빈 정의는 파라미터 이름으로 빈을 특정하고, `Step`의 빈 정의는 Tasklet의 구현 클래스를 통해 빈을 특정하게 되어 있습니다. 일관성이 없다고 느껴지는데, Tasklet은 구현 클래스에 `@Component`를 붙이기 때문이 아닌가 싶습니다.

그래서 차라리 이 Configuration에서 Tasklet에 대한 빈 정의까지 한다면, batch job 전체의 conf를 아래처럼 리팩토링 해볼 수 있을 것 같습니다.


```java
    @Bean
    public Job cacheFeedbackAvailableDowithAndLazyDowithMemberJob() {
        return new JobBuilder(JOB_NAME, jobRepository)
                .incrementer(new RunIdIncrementer())
                .start(step())
                .build();
    }

    @Bean
    @JobScope
    public Step step() {
        return new StepBuilder(STEP_NAME, jobRepository)
                .tasklet(tasklet(), platformTransactionManager)
                .build();
    }

    @Bean
    @StepScope
    public Tasklet tasklet() {
        return new CacheFeedbackAvailableDowithAndLazyDowithMemberTasklet(jdbcTemplate);
    }
```

gyun-ky replied:
> Job의 빈 정의는 파라미터 이름으로 빈을 특정하고, 

-> 메서드명으로 특정하지 않나요?

gyun-ky replied:
말씀하신 사항을 전체적으로 못알아들었씁니다,,,

gyun-ky replied:
아아 tasklet은 POJO로 납두고 아예 config에서 생성해서 빈등록을 시키자! 이거군요! 아예 한 config 파일에서 배치에 대해서 다 볼 수 있도록

S-J-Kim replied:
넵 그 의도로 위의 코드를 작성해본건데 어떻게보시는지..

gyun-ky replied:
좋은거 같습니다!

S-J-Kim committed:
fix: ranking 테이블 변경 롤백, 배치에서 DB 저장 없도록 수정

S-J-Kim committed:
fix: 배치 실행 파라미터 수정

S-J-Kim committed:
fix: 배치 스케쥴링 크론식 수정

S-J-Kim committed:
fix: 잔소리 대상 두윗 조회 조건 수정

S-J-Kim committed:
Merge branch 'develop' into feature/TAS-434

S-J-Kim committed:
fix: cache 이름 수정

S-J-Kim committed:
refactor: CachePolicy를 인터페이스 구현으로 수정

S-J-Kim committed:
feat: Redis Ops 전용 모듈

S-J-Kim committed:
fix: Redis 활용 성격에 알맞도록 이름 변경; Cache -> Store

S-J-Kim committed:
spotless

S-J-Kim committed:
fix: RedisOperator 를 제네릭을 받도록 수정

S-J-Kim committed:
feat: 잔소리 대상 두윗 & 레이지 두윗러 Redis 적재 배치

S-J-Kim committed:
refactor: 잔소리 대상 두윗 & 레이지 두윗러 Redis 적재 배치 QueryDSL로 수정

S-J-Kim committed:
feat: RedisOperator 예외 처리 추가

S-J-Kim committed:
feat: RedisOperator에 getHashes 추가

S-J-Kim committed:
fix: Redis pipeline 연산을 sessionCallback으로 수정

S-J-Kim committed:
refactor(redis): RedisOperator 메서드 일관성 및 안정성 개선 (SessionCallback, List반환)

S-J-Kim committed:
feat: 레이지 두윗러 Redis 적재를 모든 두윗의 id를 적재하는 것으로 수정 (인덱스 목적)

S-J-Kim committed:
fix: 배치 이름을 cache 이름 수정에 따라 변경;
StoreFeedbackAvailableDowithAndLazyDowithMember -> StoreFeedbackAvailableDowithTasks

S-J-Kim committed:
feat: - StoreFeedbackAvailableDowithTasksTasklet의 id 목록 적재를 Atomic rename 전략으로 수정.
- RedisOperator 에 delete 및 rename 연산 추가

S-J-Kim committed:
refactor: 배치의 Redis 접근 로직을 Repository 레이어를 통해 처리하도록 리팩토링

S-J-Kim committed:
test: 잔소리 대상 두윗 목록 적재 배치

S-J-Kim committed:
refactor: 잔소리대상 두윗 목록 조회 스펙변경; 시작일자 (date) 추가

gyun-ky commented:
지금 MR Description에 비즈니스에 대한 설명/ 데이터 흐름과 플로우가 적혀있지 않아서 파악하기 힘듭니다 ㅠㅠ

근본적으로 플로우에 대해서 여쭤봅니다

피드에 보여줄 레이지 두윗러를 골라내는 기획이 있고 비즈니스 로직이 있는데

잔소리 대상을 전체 ids를 Redis에 리스트로 store한다면 

레이지 두윗러를 계산하는 로직은 추후에 API Service에 있게 되나요???

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/common/cache/CachePolicy.java:
```diff
@@ -4,13 +4,13 @@
-    DOWITH_TASK(CacheName.DOWITH_TASK, RedisValueType.HASH, Duration.ofDays(14)),
-    LAZY_DOWTIH_TASK(CacheName.LAZY_DOWITH_TASK, RedisValueType.LIST, null);
+    DOWITH_TASK(CacheName.DOWITH_TASK, RedisValueType.HASH, Duration.ofHours(1)),
+    DOWITH_TASK_IDS(CacheName.DOWITH_TASK_IDS, RedisValueType.LIST, Duration.ofHours(1));
```
cache와 store를 구분해서 가기때문에 Store에 해당하는 것들은 따로 Policy를 분리하는 것이 좋을거 같습니다

- CachePolicy
- StorePolicy

그리고 common/cache가 아닌 common/redis로 바꿔야겠네요

gyun-ky replied:
policy enum은 명확하면 좋을거 같습니다

- DOWITH_TASK_IDS 보다는 FEED_BACK_AVAILABLE_DOWITH_TASK_IDS

근데 첫번째 코멘트로 남겼듯이 전체 플로우 먼저 설명한번 해주시면 좋을거 같습니다

S-J-Kim replied:
StorePolicy로 분리하도록 함.
- 레이지 두윗러 두윗 ID
- 잔소리 대상 두윗 목록

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/feed/cache/FeedCacheCommandRepository.java:
```diff
@@ -0,0 +1,9 @@
+import com.LetMeDoWith.LetMeDoWith.infrastructure.feed.query.dto.FeedbackAvailableDowithTaskQueryDto;
+import java.util.List;
+
+public interface FeedCacheCommandRepository {
```
DIP 의존역전의법칙 보장하려고 interface 만드신거 같은데 infrastructure 레이어에 위치해 있으면 다른 레이어가 infratructure 레이어에 의존하는 형태가 됩니다

- DIP 보장하기 위해서는 interface가 batch / domain 레이어에 있어야합니다

근데 아래 코멘트에 적겠지만 근본적으로 해당 메서드가 품고 있는 로직은 tasklet의 배치 서비스 로직이라서 

tasklet에 위치해야하는게 아닌가 싶습니다

S-J-Kim replied:
Query에 대한 repository interface는 application 레이어에 두는 것으로 결정함

gyun-ky replied:
이거 반영 안된거 같습니다

S-J-Kim replied:
cache 관련 repository는 삭제하였습니다. (캐싱하지 않기로 결정됨)

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/feed/cache/FeedCacheCommandRepositoryImpl.java:
```diff
@@ -0,0 +1,37 @@
+    private final RedisOperator redisOperator;
+
+    @Override
+    public void refreshFeedbackAvailableDowithTasks(List<FeedbackAvailableDowithTaskQueryDto> dowithTasks) {
```
여기에 구현된 건 배치의 서비스 로직인데 repository에 있는게 맞나 싶습니다

- batch 패키지에 RedisOperator interface 위치 (나중에 application 레이어에서도 쓰일 예정이니, application 레이어에도 interface 하나 만들기)
- infrastructure 레이어에는 위의 interface 구현한 RedisOperator 구현체 위치 -> DIP 보장
- tasklet에서는 redisOperator 주입 받고, 현재 메서드에 있는 로직이 tasklet에 위치

가 더 역할 책임 측면에서 적합하지 않을까 싶습니다

S-J-Kim replied:
`RedisOperator`는 인프라 영역에, 이에 대한 인터페이스 (`StoreOperator`)는 common 영역에 두고, tasklet에서는 인터페이스와 도메인 서비스를 호출하는 것으로 함.

S-J-Kim replied:
StoreOperator를 만들려다 보니, RedisOperator가 레디스 자료구조에 특화되어 작성이 되어 있어서 인터페이스를 만드는게 (단순 역방향 의존을 방지하기 위한다기엔) 크게 의미가 없다고 느껴집니다. 

제가 NoSQL 등 구현 기술의 변경에 대응해 StoreOperator가 좋겠다고 말을 했는데, 정작 Hash, ZSET 같은건 레디스에만 있으니까요..

그래서 

> - tasklet에서는 redisOperator 주입 받고, 현재 메서드에 있는 로직이 tasklet에 위치

이 구현으로 가보려 하는데 어떻게 생각하시는지.. (물론 레이지 두윗러 계산 등에서 여전히 도메인 서비스는 필요)

gyun-ky replied:
> 인터페이스를 만드는게 (단순 역방향 의존을 방지하기 위한다기엔) 크게 의미가 없다고 느껴집니다.

이게 젤 큰 의미 아닐까요??

S-J-Kim replied:
표면상 인터페이스일 뿐 역방향 의존을 방지하기 위해 구현체에서 인터페이스를 도출한거니 인터페이스를 사용하는 의도에 정확하게 반하는 설계라고 생각이 들었습니다..

그래서 저는 원래 설계처럼 CacheRepository (내지는 StoreRepository) 의 구현체에서 RedisOperator를 주입받는 방식이 DIP에 부합하지 않나 생각합니다. (잔소리 대상 두윗 같은 store 성격의 read model은 구현기술을 NoSQL 등으로 교체할 수도 있는거니, Redis에 구체적인 코드는 인프라 구현 레벨에서만 존재하는 걸로)

gyun-ky replied:
좋은 토론거리 공부거리 감사합니다 말씀해주셔서 저도 다시 살펴봤습니다

> 그래서 저는 원래 설계처럼 CacheRepository (내지는 StoreRepository) 의 구현체에서 RedisOperator를 주입받는 방식이 DIP에 부합하지 않나 생각합니다. (잔소리 대상 두윗 같은 store 성격의 read model은 구현기술을 NoSQL 등으로 교체할 수도 있는거니, Redis에 구체적인 코드는 인프라 구현 레벨에서만 존재하는 걸로)

-> 근데 이렇게 하면 현재 저희 비즈니스에 비해서 너무 코드 복잡도가 많아진다고 생각 들었습니다. (현재도 저희 비즈니스에 비해 제가 욕심 부려서,,, DDD도입하면서 너무 복잡해졌나,, 기능 개발 속도가 안나온다,,이런 느낌이라) 단순히 Redis를 사용하는 건데 Repository 파일도 만들고 너무 복잡도가 많아져서요

그리고 만약 StoreRepository의 구현 기술이 NoSQL로 변경되는 경우
- 현재는 RedisOperator 메서드 이름이 몇몇개의 Redis에 핏하게 만들어진 메서드명 빼고는 사실상 NoSQL로도 충분히 구현 가능한 자료구조 action 명입니다
- MongoDB를 사용한다 가정했을 때,MongoDBRepository (예시) 구현체안에서는 MongoDB sql을 사용해서 pushRightAll 같은 메서드를 구현하면 됩니다
=> 그래서 말씀하신 그런 구현체가 변경될걸 고려하면 RedisOperator의 메서드 명들을 자료구조 store 관점에서 수정하는게 좋지 않을까 싶습니다


> 표면상 인터페이스일 뿐 역방향 의존을 방지하기 위해 구현체에서 인터페이스를 도출한거니 인터페이스를 사용하는 의도에 정확하게 반하는 설계라고 생각이 들었습니다..

위에 말씀드린 해결방법이 의도에 반하지 않게 만들거라고 생각하고요. 레이어 아키텍처에서는 어디에선 DIP되어있고 어디에선 예외를 두어서는 안된다고 생각합니다.

S-J-Kim replied:
저는 Repository를 쓰는게 복잡도가 많아진다는 부분이 공감이 어려운 것 같습니다.

일반적인 조회 유스케이스를 봐도 QueryRepository <- QueryRepositoryImpl <- JPAQueryFactory 순서로 의존을 하게 되는데 이 경우는 JPAQueryFactory가 아니고 RedisOperator 일 뿐이라고 보거든요..

또 더해서, 구현체가 변경될 가능성이 있으면 더욱이 Repository로 감싸야 한다고 봅니다. Redis와 NoSQL에 항상 물리적으로 동일한 구조로 데이터를 저장할 수는 없을 것이기 때문입니다. 애플리케이션 레벨에서 활용하는 데이터의 스키마는 같아도 어디선 통으로 조회를 할 수도 있고 어디선 조인을 해서 가져와야 할 수도 있는거니까요.. (혹은 Redis와 NoSQL이 상호 대비되는 관계가 아니고 두개가 모두 필요한 유스케이스도 필요할 수도 있겠습니다,, 물론 지금으로선 논외)

그래서 Operator는 인프라 영역에서 각 구현기술에 최대한 fit하게 만들어 두고, Repository 레벨에서 Service에 줘야하는 데이터의 형태에 맞춰 구현기술의 operation을 조합해서 가져오는 패턴이 '추상화에 의존한다'는 관점에 부합하지 않나 싶습니다.

gyun-ky replied:
요건 실제 코드 예시를 하나 만들어서 일요일에 시간되시면 온라인으로 한번 얘기하시는건 어떠할까유??

gyun-ky replied:
온라인 회의 의사결정
- Store 관점에서는 Repository로 캡슐화해서 상요
- RedisOperator는 Repository impl에서만 사용

S-J-Kim replied:
삭제되었습니다..

S-J-Kim commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
Cache 로서의 레디스와, Store로서의 레디스의 활용 방법을 분리하는 관점에서 이 클래스를 작성했습니다.

캐시 용도는 Cache Manager를 적극 활용하고, store 관점의 명령들은 이걸 통해서 작성하는 것이 어떨까 싶습니다.

그렇다면 CacheHelper가 지금보다 더 간단하게 작성이 가능할 것 같습니다.

S-J-Kim committed:
chore: 디렉토리 이름 변경; common.cache -> common.redis

S-J-Kim committed:
chore: 이름 변경;
- CachePolicySpec -> RedisPolicySpec
- CacheName -> keyName;

S-J-Kim committed:
feat: store 목적의 key는 StorePolicy로 분리

S-J-Kim committed:
refactor: Redis store key 이름 변경
- DOWITH_TASKS -> FEEDBACK_AVAILABLE_DOWITH_TASKS
- DOWITH_TASK_IDS -> LAZY_DOWITH_TASK_IDS

S-J-Kim committed:
feat: feed 의 Dowith 도메인 모델 및 레이지 두윗러 추출 도메인 서비스

S-J-Kim committed:
rename: feedback-available 단어 삭제 -> 조금 더 universal 한 단어로 수정함

S-J-Kim committed:
feat: tasklet에서 repository 주입받지 않고 직접 RedisOperator 사용하도록 수정

S-J-Kim commented on src/main/java/com/LetMeDoWith/LetMeDoWith/batch/tasklet/StoreFeedbackAvailableDowithTasksTasklet.java:
```diff
@@ -0,0 +1,83 @@
+        List<FeedDowithTask> lazyDowithTasks =
+                lazyDowithTaskSelector.selectLazyDowithTasks(feedDowithTasks, LAZY_DOWITH_COUNT);
+        List<String> lazyDowithTaskIds =
+                lazyDowithTasks.stream().map(dowith -> dowith.id().toString()).toList();
```
잔소리 가능 두윗 중에서 레이지 두윗을 선별하는 로직은 도메인 서비스로 분리했습니다.

S-J-Kim commented on src/main/java/com/LetMeDoWith/LetMeDoWith/batch/tasklet/StoreFeedbackAvailableDowithTasksTasklet.java:
```diff
@@ -0,0 +1,83 @@
+            String tempKeySuffix = "temp_" + SystemTimeUtil.nowTime();
+            redisOperator.pushRightAll(StorePolicy.LAZY_DOWITH_TASK_IDS, tempKeySuffix, lazyDowithTaskIds);
+            redisOperator.rename(StorePolicy.LAZY_DOWITH_TASK_IDS, tempKeySuffix, "");
+        }
```
레이지 두윗러의 경우 DELETE -> RPUSH 하는 것이 아니고,
1. 임시 키에 모두 RPUSH
2. 임시 키의 이름을 레이지 두윗러의 원래 키 이름으로 RENAME

하는 방식으로 작성했습니다.

이 방법이 LIST 전체를 원자적으로 교체하기 때문에 더 안전한 방법이라고 합니다.

S-J-Kim replied:
또한 현재는 레이지 두윗러의 경우 Top 15만 추출해서 적재하는데, 잔소리 대상 두윗의 전체 Id를 레이지 두윗러로 레디스에 적재하는 방법을 제안합니다. (갯수를 15개 -> 전체로 늘리자는 말)

아래의 이유로 이득이 있다고 봅니다.

잔소리 대상 두윗을 전체 조회하기 위해선 우선 HSCAN으로 키를 모두 조회 후, 이 키 목록으로 HASH인 키들을 조회 해와야 함.
하지만 SCAN 명령어는 아래의 불편함 존재
- 항상 조회 할 갯수를 지정해 주어야 함 (조회 시점에 전체 갯수를 알 수 없고, 무작정 큰 수로 설정할 수 도 없음)
- 심지어 이 갯수대로 조회함을 보장 하지 않음 ([Redis docs에 언급](https://redis.io/docs/latest/commands/scan/#scan-guarantees)되어 있습니다.)
- 갯수가 많아 다음 커서까지 필요하다면, 조회 1번에 키 가져오기 불가능. (커서에 대한 내용은 [Redis의 SCAN 문서](https://redis.io/docs/latest/commands/scan/)를 참고하세요)

-> 따라서 레이지 두윗러 id 목록을 전체 잔소리 대상 두윗에 대한 id를 저장하면 이 목록을 가져와 HGETALL에 던지기만 하면 됩니다.
-> 심지어 도메인 서비스의 로직이 전체 잔소리 대상 두윗에 대해 레이지 두윗러 순위를 계산하고 Top 15만 가져가고 있기 때문에 전체 id를 저장한다면 버려지는 데이터가 없습니다.
-> Top 15개 리스트가 5개 미만으로 줄어들 경우 배치를 수동으로 트리거해야 하는데, 전체 id 리스트를 인덱싱 할 경우 5개 미만으로 줄어들 가능성이 대폭 줄어듭니다.

gyun-ky replied:
> Redis 임시키에 RPUSH -> RENAME 관련

일단 두개의 명령어가 각각 Request, 총 2개의 Request로 진행되기 때문에 Redis 입장에서는 원자적이라고 볼수 없는거 같습니다

하지만 말씀하신대로 WAS 입장에서는 LIST가 교체되는 중간과정은 볼수 없으니 최종상태 측면에서는 원자적이라고 볼수 있을거 같습니다

하지만, RENAME전에 프로세스가 죽는다면 tempKey만 영구히 남겨있게 되는 상황이 생길 수 있으므로

```
RPUSH tempKey a b c
EXPIRE tempKey 30

RENAME tempKey realKey
PERSIST realKey
```
와 같이 진행하는게 더 안전하지 않을까 싶습니다

그리고 이 로직을 redisOperator 안에 두고 RedisOperator로 추상화해서 쓰는건 어떠할까요??

아예 최종적 원자성을 가져가기 위해 
- MULTI - EXEC 
- lua script
로 작성하여 Redis 입장에서도 원자성 확보하는 것도 좋은 방법일거 같습니다

gyun-ky replied:
그리고 저희같은 피드 비즈니스에는
* 버전드 스냅샷 저장소
방식이 더 적합할 수있겠다는 생각이 들긴했는데 (Dowith 완료 처리(reader)도 고려시에,,,)

일단은 시간이 없으니 Redis에 적재하는 방식을 바꾸는 형식으로 가진 않아도 될거 같습니다

gyun-ky replied:
> 잔소리 대상 두윗을 전체 조회하기 위해선 우선 HSCAN으로 키를 모두 조회 후, 이 키 목록으로 HASH인 키들을 조회 해와야 함.
> 하지만 SCAN 명령어는 아래의 불편함 존재
> 
> 항상 조회 할 갯수를 지정해 주어야 함 (조회 시점에 전체 갯수를 알 수 없고, 무작정 큰 수로 설정할 수 도 없음)
> 심지어 이 갯수대로 조회함을 보장 하지 않음 ([Redis docs에 언급](https://redis.io/docs/latest/commands/scan/#scan-guarantees)되어 있습니다.)
> 갯수가 많아 다음 커서까지 필요하다면, 조회 1번에 키 가져오기 불가능. (커서에 대한 내용은 [Redis의 SCAN 문서](https://redis.io/docs/latest/commands/scan/)를 참고하세요)
> -> 따라서 레이지 두윗러 id 목록을 전체 잔소리 대상 두윗에 대한 id를 저장하면 이 목록을 가져와 HGETALL에 던지기만 하면 됩니다.
> -> 심지어 도메인 서비스의 로직이 전체 잔소리 대상 두윗에 대해 레이지 두윗러 순위를 계산하고 Top 15만 가져가고 있기 때문에 전체 id를 저장한다면 버려지는 데이터가 없습니다.

이 부분이 저희 현재 요구사항과 관련있는지 잘 이해가 되지 않습니다

* 기존 LAZY DOWITH 15개만 캐싱 -> 잔소리 가능한 Dowith을 전체 캐싱하는 것은 오케이
* 조회 시에는 개수를 지정해준다는게 이해가 안됩니다 -> DB에서 dowithTaskId들 먼저 가져온 후에 dowithTaskId 들로 조회하지 않나요?

S-J-Kim replied:
조회시 갯수 지정 -> Redis SCAN 명령어를 사용할 때 스캔 할 갯수를 파라미터로 넘겨야 합니다. (아니면 기본값이 10입니다)

> DB에서 dowithTaskId들 먼저 가져온 후에 dowithTaskId 들로 조회하지 않나요?

이게 배치의 로직을 말하는건지 API의 로직을 말하는건지 헷갈리는데, 배치 로직은 description의 sequence diagram 참고하시면 되고, API에서는 Db에서 id를 조회하는 방향으로 가지 않으려 합니다. (그럼 Redis를 쓰는 이유가 없으니까)

LAZY_DOWITH 키에 잔소리 가능한 두윗의 모든 id를 넣으면 API 로직을 이렇게 할 수 있습니다.

```
LRANGE LAZYDOWITH 0 -1
HGETALL DOWITHTASK::{id} (LAZYDOWITH 길이만큼 pipeline)
```

S-J-Kim replied:
> > Redis 임시키에 RPUSH -> RENAME 관련
> 
> 일단 두개의 명령어가 각각 Request, 총 2개의 Request로 진행되기 때문에 Redis 입장에서는 원자적이라고 볼수 없는거 같습니다
> 
> 하지만 말씀하신대로 WAS 입장에서는 LIST가 교체되는 중간과정은 볼수 없으니 최종상태 측면에서는 원자적이라고 볼수 있을거 같습니다
> 
> 하지만, RENAME전에 프로세스가 죽는다면 tempKey만 영구히 남겨있게 되는 상황이 생길 수 있으므로
> 
> ```
> RPUSH tempKey a b c
> EXPIRE tempKey 30
> 
> RENAME tempKey realKey
> PERSIST realKey
> ```
> 
> 와 같이 진행하는게 더 안전하지 않을까 싶습니다
> 
> 그리고 이 로직을 redisOperator 안에 두고 RedisOperator로 추상화해서 쓰는건 어떠할까요??
> 
> 아예 최종적 원자성을 가져가기 위해
> 
> * MULTI - EXEC
> * lua script
>   로 작성하여 Redis 입장에서도 원자성 확보하는 것도 좋은 방법일거 같습니다

`RedisOperator` 에서 기본적으로 `StorePolicy` 에 정의된 ttl을 명령 실행 때 동시에 적용하고 있기 때문에 tempKey만 영구히 살아있는 상황은 발생하지 않습니다.

gyun-ky replied:
> RedisOperator 에서 기본적으로 StorePolicy 에 정의된 ttl을 명령 실행 때 동시에 적용하고 있기 때문에 tempKey만 영구히 살아있는 상황은 발생하지 않습니다.

아 확인했습니다 근데 생각을 해보니 근본적으로 확인 필요한게
- Store는 기본적으로 TTL을 걸면 안되지 않을까 싶습니다. 애초에 Cache가 아니고 DB처럼 데이터 적재를 사용하는거기 때문에 데이터 생명주기를 서버가 컨트롤 해야한다고 생각합니다

그런차원에서 StorePolicy에 대한 근간 정책이 변경되어야할거 같습니다

gyun-ky replied:
- DB에 LAZY_DOWITH 리스트 따로 적재하지 않는것 확인했습니다

Feed 레이지 두윗러 리스트 조회 API 차원에서 코멘트 다신거로 이해해서 질문드렸었고

말씀하신대로 해당 API에서 이렇게 조회한다고 말씀해주셨는데
```
LRANGE LAZYDOWITH 0 -1
HGETALL DOWITHTASK::{id} (LAZYDOWITH 길이만큼 pipeline)
```
- 요구사항은 5개를 레이지 두윗 5개 Feed에 보여줌
- 도중에 레이지 두윗이 완료되면 Dowith완료 API에서 LAZY_DOWITH List에서 해당 id 제거

위의 플로우가 조회의 주요 요구사항이고

LAZY_DOWITH List에서 앞에 5개만 가져오고 5개만 캐싱된 Dowith을 조회하면 되는데

왜 전체를 조회해서 HGETALL하는 이유는 무엇인가요?

S-J-Kim replied:
> Store는 기본적으로 TTL을 걸면 안되지 않을까 싶습니다. 애초에 Cache가 아니고 DB처럼 데이터 적재를 사용하는거기 때문에 데이터 생명주기를 서버가 컨트롤 해야한다고 생각합니다

저도 store는 persistence 성격이 있으니 ttl은 조심스럽게 접근해야 한다고 생각합니다. 다만, 이 데이터는 비즈니스적으로 유효한 시간이 정해져있기 때문에 이 시간을 지나서 조회되면 요구사항에 맞지 않는 데이터가 조회되는 것 입니다. 

만약 불의의 이유로 배치가 한번 수행되지 않았다면, stale data가 조회 (즉 비즈니스적으로 옳지 않은 결과) 될 수 있습니다. 만약 ttl이 있다면 expire되어 fallback 로직이 수행되기 때문에 속도는 redis 조회 대비 느릴지언정 틀린 데이터가 조회되지는 않습니다.

이 의도로 이 데이터는 ttl이 존재하는것이 유리하다고 생각합니다. 어떻게 보시는지?

S-J-Kim replied:
> DB에 LAZY_DOWITH 리스트 따로 적재하지 않는것 확인했습니다
> Feed 레이지 두윗러 리스트 조회 API 차원에서 코멘트 다신거로 이해해서 질문드렸었고
> 
> 말씀하신대로 해당 API에서 이렇게 조회한다고 말씀해주셨는데
> 
> LRANGE LAZYDOWITH 0 -1
> HGETALL DOWITHTASK::{id} (LAZYDOWITH 길이만큼 pipeline)
> 요구사항은 5개를 레이지 두윗 5개 Feed에 보여줌
> 도중에 레이지 두윗이 완료되면 Dowith완료 API에서 LAZY_DOWITH List에서 해당 id 제거
> 위의 플로우가 조회의 주요 요구사항이고
> 
> LAZY_DOWITH List에서 앞에 5개만 가져오고 5개만 캐싱된 Dowith을 조회하면 되는데
> 
> 왜 전체를 조회해서 HGETALL하는 이유는 무엇인가요?

아 저는 LAZY_DOWITH 이 "모든 잔소리 가능 두윗"의 id 를 가지고 있다면 "레이지 두윗러 리스트 조회 API"는 따로 필요 없다는 취지에서 말한 부분입니다.

레이지 두윗러는 잔소리 가능 두윗의 부분집합임은 이미 알고계실겁니다.

따라서 (LAZY_DOWITH 이 "모든 잔소리 가능 두윗"의 id 를 가지고 있다면) 각각의 유스케이스를 아래처럼 만들 수 있습니다.

```
// 잔소리 대상 두윗 조회
LRANGE LAZYDOWITH 0 -1
HGETALL DOWITHTASK::{id} (LAZYDOWITH 길이만큼 pipeline)

// 레이지 두윗러 조회
LRANGE LAZYDOWITH 0 5
HGETALL DOWITHTASK::{id} (5개 pipeline)
```

그리고 여기서 더 디벨롭해본다면, 화면에서 잔소리 대상 두윗 조회 API만 호출한다음 앞에 5개를 레이지 두윗러 zone에서 보여주도록 하면 됩니다. (잔소리 대상 두윗 조회 API의 결과는 레이지 두윗러 순위에 맞춰 정렬되어 있기 때문)

gyun-ky replied:
말씀하시는 건 그러면
- Feed UI상 두개의 섹션 (레이지 두윗러, 잔소리 대상 목록) 을 한개의 API로 서빙
- 앱에서 한개의 API 호출해서 잘라서 쓴다
는 말씀으로 이해했습니다만 위험한 설계이지 않을까 싶습니다

아래 사항 고려하시면 API를 따로 빼는게 적절할거 같습니다
- 잔소리 대상 두윗 조회 정렬 기준이 바뀐다면?
- 잔소리 대상 두윗 조회 Paging 처리

S-J-Kim replied:
~말씀하신 내용을 고려해보니 API를 분리하는 것이 맞을 것 같습니다~

~`FEEDBACK_AVAIL_DOWITH::latest` (최신순)~
~`FEEDBACK_AVAIL_DOWITH::feedbackCount` (잔소리 많은순 - 예시)~

~이런식으로 id list를 기준에 따라 정렬된 상태로 가져오고 hash를 조회하는 방식으로 api 를 만들겠습니다.~

https://github.com/LetMeDoWith/LetMeDoWith-BE/pull/73#discussion_r2698765291 이 코멘트 얘기한대로 배치 아예 안쓰는 방법을 말씀드려봅니다.

S-J-Kim replied:
배치를 사용하지 않기로 결정하여 삭제되었습니다

S-J-Kim commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/feed/cache/FeedCacheCommandRepositoryImpl.java:
배치에서 `RedisOperator`를 직접 주입받는다면 이 클래스는 삭제 예정입니다.

gyun-ky replied:
삭제 조치 안된거 같습니다!

S-J-Kim replied:
삭제 예고 였습니다... 이제 삭제하겠습니다..ㅋ

gyun-ky replied:
조치되었으면 코멘트 주세요!

S-J-Kim replied:
삭제되었습니다

S-J-Kim commented on src/main/java/com/LetMeDoWith/LetMeDoWith/domain/feed/service/LazyDowithTaskSelector.java:
```diff
@@ -0,0 +1,18 @@
+                        .thenComparing(FeedDowithTask::feedbackCount, Comparator.reverseOrder()))
+                .limit(size)
+                .toList();
+    }
```
도메인 서비스에서 시간과 `feedbackCount`에 대한 정렬을 모두 수행하는데,

현재 `DowithTask` 의 테이블 구조상 시작 시간이 `DateTime`의 형태가 아니고, `feedbackCount` 또한 서브쿼리로 가져오기 때문에 인덱스를 탈 수 없기 때문입니다.

S-J-Kim reviewed (commented):
1차 리뷰 이후 수정사항에 대한 설명입니다. 코멘트 달린 파일 위주로 리뷰하시면 편하고 빠릅니다.
@gyun-ky

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/batch/tasklet/StoreFeedbackAvailableDowithTasksTasklet.java:
```diff
@@ -0,0 +1,83 @@
+        List<FeedDowithTaskQueryDto> dowithTaskQueryDtos =
+                feedQueryRepository.getFeedbackAvailableDowithTasks(targetDateTime);
+
+        List<FeedDowithTask> feedDowithTasks = dowithTaskQueryDtos.stream()
```
변수명을 feedbackAvailDowithTasks가 더 가독성 좋을거 같습니다

S-J-Kim replied:
배치를 사용하지 않는 것으로 결정하여 삭제되었습니다

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/batch/tasklet/StoreFeedbackAvailableDowithTasksTasklet.java:
```diff
@@ -0,0 +1,83 @@
+@Component
+@StepScope
+@RequiredArgsConstructor
+public class StoreFeedbackAvailableDowithTasksTasklet implements Tasklet {
```
Tasklet 명 헷갈리지 않게 LazyDowith을 골라낸다는 의미를 가진 이름으로 변경되면 좋을거 같습니다

S-J-Kim replied:
조금 고민이 되는게, 잔소리 대상 두윗을 선택하면서 동시에 레이지 두윗도 골라내는데, 이걸 전부 포함하도록 이름을 지어야 할지 어떻게 할지 참 난감하네요

gyun-ky replied:
지금 든 생각인데 Tasklet을 두개로 나누고, Job에서 두개의 Tasklet을 순차적으로 실행하는게 Spring Batch의 아키텍처를 잘 활용하는 방법이라는 생각도 드네요

그러면 Tasklet이 아닌 Step을 사용하는게 맞는거 같다는 생각도 들고요

S-J-Kim replied:
> 그러면 Tasklet이 아닌 Step을 사용하는게 맞는거 같다는 생각도 들고요

이게 무슨 말인지 이해가 잘 안되는군요.. Step은 Tasklet 이거나 Chunk-driven (reader-processor-writer) 둘중 하나인거 아닌가요..?

잔소리 대상 두윗 적재 step 과 레이지 두윗러 적재 step을 나누자는 의미로 이해를 했는데, 잔소리 대상 두윗과 레이지 두윗러는 데이터가 같은 곳에서 오기 때문에 step을 나누게 되면 같은 데이터를 두번 조회하게 될 것 같습니다.

gyun-ky replied:
아 잠시 STEP에 대해서는 헷갈렸습니다! 

> 그러면 Tasklet이 아닌 Step을 사용하는게 맞는거 같다는 생각도 들고요

-> 이 코멘트는 취소할게요


Tasklet을 두개로 분리하고 "잔소리 대상 두윗 적재 step 과 레이지 두윗러 적재 step을 나누자" 가 맞고요

- Step 구성을 잘 해서 잔소리 대상 두윗적재 Tasklet에서 나온 결과를 -> 레이지 두윗러 적재 Tasklet으로 전달 하는 방식

하지만 Tasklet의 독립성을 생각하면 아예 현재 설계하신 방안과 새로운 방법을 생각해보면 아래와 같이 Tasklet 두개의 설계로 갈 수도 있을거 같습니다 (뭔가 계속 꼬이고 이상하다면 기존의 설계방식을 잠깐 내려두고 제 3의 방법을 생각하는게 더 나은거 같습니다)
- Tasklet1 : DB에서 조회 -> 잔소리 대상 두윗 Redis Cache
- Tasklet2 : (Tasklet1의 결과 쓰지 않고) DB에서 SQL로 레이지두윗러 산출 계산 진행 -> id들 Redis List에 Store

S-J-Kim replied:
~~레이지 두윗러가 잔소리 대상 두윗과 독립적인 데이터라면 그렇게 나누는것이 합리적이 겠지만, 지금 기획 상으로는 레이지 두윗러는 잔소리 대상의 일부분을 추출해서 보여주는 것이기 때문에 분리함으로서 얻는 이득이 크지 않다고 생각합니다..~~

~~그리고 만약 step을 나눠서 각각 적절하게 이름을 붙였다면, 그 step들을 실행할 job은 어떻게 이름을 지어야 하는거죠....?~~

https://github.com/LetMeDoWith/LetMeDoWith-BE/pull/73#discussion_r2698765291 이 코멘트 얘기한대로 배치 아예 안쓰는 방법을 말씀드려봅니다.

S-J-Kim replied:
배치를 사용하지 않는 것으로 결정하여 삭제되었습니다

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/batch/tasklet/StoreFeedbackAvailableDowithTasksTasklet.java:
```diff
@@ -0,0 +1,83 @@
+                lazyDowithTasks.stream().map(dowith -> dowith.id().toString()).toList();
+
+        // 잔소리 대상 두윗 상세 정보 Redis 적재
+        redisOperator.putHashes(StorePolicy.FEEDBACK_AVAILABLE_DOWITH_TASKS, feedDowithTasks, dto -> dto.id()
```
잔소리 대상 두윗 상세 정보를 적재하는 건 캐싱이니까 
- cacheHelper를 사용하는게 좋지 않을까 싶습니다

그리고 Redis만 떼고 본다면 결국 DowthTask에 대해서 캐싱하는 것이니 
- CachePolicy.DOWITH_TASK 로 `dowith-task::{dowithTaskId}` 를 key로 가지고 캐싱하는건 어떠할까요??

S-J-Kim replied:
캐시와 스토어는 분리하자고 하던거 아니던가요...

그리고 잔소리 대상 두윗 정보는 값이 바뀔 수 있으니까 (두윗 완료/삭제 시) 캐시는 아니라고 보았습니다.

gyun-ky replied:
캐시와 스토어 분리 맞습니다

근데 말씀하신 DowithTask를 적재하는건 캐싱이라고 봤습니다

> 잔소리 대상 두윗 정보는 값이 바뀔 수 있으니까 (두윗 완료/삭제 시) 캐시는 아니라고 보았습니다.

-> 이게 어떤 말씀이신지요? 캐시란 언제든 값이 바뀔가능성을 열어두고 값이 바뀔때 cache도 업데이트해주어야하는거 아닌가요??
잔소리 대상 두윗 정보값이 바뀌면 update API에서 캐시를 업데이트해주는게 맞지 않나요??

S-J-Kim replied:
데이터의 일부분을 수정하는데 캐시라고 볼 수 있나요? 저는 Spring Cache 사상을 넘어서는 사용 케이스는 store로 봐야 한다고 봅니다. (하나의 오브젝트를 온전하게 캐싱하고, 변경시는 덮어쓰기)

gyun-ky replied:
> 저는 Spring Cache 사상을 넘어서는 사용 케이스는 store로 봐야 한다고 봅니다.

저는 Spring Cache는 메서드 단위의 캐싱하는 기술정도로 생각하고 있습니다

저희가 추상화한 cache는 정말 자주 사용되는 데이터들을 빠르게 조회하기 위한 Cache 라는 논리적 개념으로 보고 있습니다

> 데이터의 일부분을 수정하는데 캐시라고 볼 수 있나요?

그런 측면으로는 데이터를 일부분 수정한다랑 cache랑 관련이 크게 있지는 않고
- 비즈니스 플로우 측면에서 이 데이터에 대한 접근이 자주 일어나고
- RTT가 늘어나 UX에 부정적 영향을 주는 것
에 대해 좀 더 초점 두고 있긴합니다

그런 측면에서 DowithTask는 캐싱 대상이라고 생각하고 그에 대한 비즈니스적인 이유는 다른 쓰레드에 좀 더 자세히 적었습니다

S-J-Kim replied:
저는 둘러보기에서 조회되는 데이터가 베치를 통해 pre-computed 되어 redis 에 올라가는 관점에서 store 라고 말씀을 드리고 있는건데, 캐싱의 관점으로 두윗을 Redis에 올린다고 하면 차라리 CacheManager가 제공하는 애너테이션을 써서 API 레벨로 캐싱 해버리는게 나을 것 같습니다. (페이징 고려하면 repo 레이어에 캐싱도 가능)

TTL은 배치 주기보다 훨씬 짧게 해서 (1분 이내) 두윗 완료/삭제 api에 캐시 조작 로직을 넣지 말고, 그냥 캐시가 일시적으로 정합성을 보장하지 않음을 인정하고 구현을 최대한 간단하게 가는 방향으로 가는게 생산성 면에서 나을 것 같습니다. (나중에 이게 UX에 크게 영향이 생긴다 싶으면 지금 구조로 롤백하시죠)

이렇게 하면 배치도 따로 필요 없을 것 같고, 잔소리 대상 두윗도 그냥 직렬화 해서 string으로 넣어버리면 조회도 pipeline을 안해도 되니 좋다는 생각입니다.

S-J-Kim replied:
둘러보기 설계가 너무 진척이 안나오는데, 제가 cache와 store에 대한 명확한 기준을 제시하지 못하기도 했고, 현재 설계 자체가 cache와 store의 특성을 모두 가지고 있기 때문이라 생각합니다.

RedisOperator 모듈 자체는 후에 쓸모가 있을테니 유지하고 feed API들은 아예 Spring Cache 레벨에서 캐싱하는 걸로 가시죠.

gyun-ky replied:
> CacheManager가 제공하는 애너테이션을 써서

캐싱을 꼭 애너테이션을 사용해서 하기보다는 애너테이션으로 어려운 부분들은 cacheHelper로 하는 방법을 초기에 생각하긴했었는데요

지금 사실 로직은 다 완성되었고 잘 개발하셨는데요

Cache와 Store에 대한 정의가 지금 저희가 엇갈려서 공통이 안잡혀서 계류하고 있는거로 느껴집니다

Store와 Cache에 대해서 다른 레퍼 상호 많이 찾아보고

일요일에 온라인으로 구두로 대화하는게 어떠할까요??

S-J-Kim replied:
네 저도 "애너테이션을 쓴다" 라고 표현했지만, Policy에 통합되어 있는 CacheHelper 로 모든 캐시 조작이 이루어지는게 맞다고 생각합니다.

캐시와 스토어에 대한 정확한 정의와 구분은 해봐야겠지만, 지금 이 유스케이스는 확실히 캐시라고 하기도, 확실히 스토어라고 하기도 어렵다는 생각을 계속 하고 있습니다.

그런 관점에서 cache/store에 대한 정확한 구분을 정의하는 것과 동시에 feed의 api에 대한 부분은 배치가 꼭 필요한가에 대해서도 얘기했으면 합니다.

S-J-Kim replied:
배치를 사용하지 않는 것으로 결정하여 삭제되었습니다

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+    // Key Ops
+    // ==================================================================================
+
+    public void delete(RedisPolicySpec policy, String key) {
```
key에 빈 문자열을 전달하게 하기 보다는

null을 허용하고 buildKey에서 key에 null이 들어오면 `prefix::` -> `prefix`로 key를 생성하도록 하는게 좋을거 같습니다

gyun-ky replied:
redisOperator.delete(StorePolicy.LAZY_DOWITH_TASK_IDS, null)

S-J-Kim replied:
이미 그렇게 되어있습니다.

```java
    private String buildKey(RedisPolicySpec policy, String key) {
        if (key == null || key.isBlank()) {
            return policy.keyName();
        }
        return policy.keyName() + "::" + key;
    }
```

gyun-ky replied:
사용하시는 클라이언트 코드에서는 

redisOperator.delete(StorePolicy.LAZY_DOWITH_TASK_IDS, "")

이렇게 사용하고 계십니다!

S-J-Kim replied:
제가 어느 부분을 놓치고 있는지 잘 모르겠습니다만...

`redisOperator.delete(StorePolicy.LAZY_DOWITH_TASK_IDS, "")` 를 수행할 때 `delete` 메서드 안의 buildKey에는 `(StorePolicy.LAZY_DOWITH_TASK_IDS, "")` 가 파라미터로 넘어갈 것이고, `"".isBlank()` 는 `true` 입니다.

gyun-ky replied:
공통 혹은 라이브러리 개발 시에 해당 공통 사용하는 개발자는 메서드 스키마만 보고 사용할건데요

키 설계가
- {prefix}::{key} 이렇게 들어간다는 것이 정책이라면

저는 redisOperator buildKey 메서드를 보지 않고, 빈 스트링을 넘기면
- {prefix}:: -> 이렇게 될거라고 추측합니다
- 현실은 {prefix} 이렇게 되지만요

RedisOperator를 개발하셨으니 buildKey 안에 메서드를 아시지만, 사용할 다른 개발자 입장에서는 해당 메서드는 보지 않고 key 설계 정책과 메서드 스키마만 보고 개발을 할 것이기 때문에 유져 개발자 입장에서 한번 검토했을때 제가 위에 드린 말씀이 맞는지 한번 확인 부탁드리겠습니다~!

S-J-Kim replied:
`::` 라는걸 separator로서 약속한 것인데 이거 뒤에 아무것도 없는 `{prefix}::` 같은 형태의 키가 존재한다고 생각할까요? 오리혀 저라면 라이브러리 레벨에서 이런 끄트머리는 알아서 최적화 해주겠거니 생각할거라는 의도입니다.

같은 관점에서 [여기 코드](https://github.com/LetMeDoWith/LetMeDoWith-BE/blob/a8d5b27138a1ad3426f1be9cad01e02aa1b70699/src/main/java/com/LetMeDoWith/LetMeDoWith/common/cache/CacheHelper.java#L143) 를 작성한거 아닌가요...

gyun-ky replied:
아 저 메서드 AI가 해줬는데 검토를 제가 못한거 같네요,,! 제가 검토를 제대로했다면 isBlank는 없앴을거 같습니다 java에서 ""가 의미있는 String으로 간주되는만큼

> :: 라는걸 separator로서 약속한 것인데 이거 뒤에 아무것도 없는 {prefix}:: 같은 형태의 키가 존재한다고 생각할까요?

저는 저게 seperator라서 더 그렇게 된다고 생각했습니다..! 알아서 최적화..?의 범주가 사실 되게 저는 애매하다고 생각해서요,,! 공통은 Strict하고 명확해야하지 않나가 제 개인적 생각입니다

S-J-Kim replied:
키 이름 규칙을 어떻게 정했느냐에 따라 다를 것 같습니다. 만약 `{prefix}::` 이 `{prefix}` 랑 별개로 의미가 있는 값이라면 이 코드는 잘못된게 맞지만 `{prefix}::` 같은 형태의 키가 없어야 한다면 개발자가 실수로 "" 로 보내더라도 redis에서 `{prefix}` 로 검색되도록 하는 게 맞는거 아닌가 싶네요

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/common/redis/CacheName.java:
```diff
@@ -1,11 +1,11 @@
     public static final String KAKAO_PUBLIC_KEY = "kakao-public-key";
-    public static final String DOWITH_TASK = "dowith";
-    public static final String LAZY_DOWITH_TASK = "lazy-dowith";
+    public static final String FEEDBACK_AVAILABLE_DOWITH_TASKS = "feed::dowithTask";
```
캐시한 DowithTask는 피드 뿐만 아니라 여러 UseCase(ex. 두윗 목록 조회, 두윗 상세 조회 등등)에서 사용될 수 있습니다. Redis에 캐싱시에 특정 UseCase 기준으로 Key설계가 아닌 데이터 관점에서 되어야할거 같습니다

- FEEDBACK_AVAILABLE_DOWTH_TASK -> DOWITH_TASK
- feed::dowithTask -> dowithTask

S-J-Kim replied:
바로 아래 코멘트처럼 이건 둘러보기 전용으로 스키마를 설계했으니 `StoreName` 등으로 별도 관리하는게 좋을 것 같습니다.

별개로 두윗 목록 조회, 상세 조회를 위해 두윗을 캐싱해야 하는지는 물음표입니다.. 둘러보기가 아니면 거의 대부분 캘린더에서 본인이 생성한 두윗 위주로 조회할테니까요..

gyun-ky replied:
> 본인이 생성한 두윗 위주로 조회할테니까요.. 

-> 이때도 유져 입장에서 딜레이가 있다면 캐싱하는게 좋지 않을까요? 예를 들어, 잔소리 대상으로 분류되어 캐싱된 Dowith들은 유져 플로우상 잔소리를 받으면 상세 조회를 많이 할거라 hit rate도 높을거 같습니다

위의 케이스처럼 언제 어디서 딜레이가 발생해서 유져의 경험이 떨어질 수 있으니 그때 바로 대응할 수 있도록 (다시 캐시 키 따고 비슷한 데이터들을 다시 적재하기보다는)

캐싱 시에 특정 use case에 얽매여서 key 설계하여 캐싱하지 말자는게 제가 말씀드린 포인트긴합니다!

S-J-Kim replied:
우선 저는 store와 cache는 분리해서 관리해야 한다고 봅니다. store라는게 read model의 성격을 지니는데, 조회 비용이 비싼 유스케이스의 성능을 높이기 위한 목적이 강하고 그 관점에서 속도가 빠른 저장소를 쓰는 것이기 때문에, store는 universal 하게 설계 할 부분은 아니라고 봅니다.


또한 잔소리 대상 두윗이 캐싱된다고 하면, 두윗 단일 조회 API의 로직에 우선 redis 조회하도록 추가해야 하는데, 잔소리 대상 두윗을 상세조회 할 사람은 두윗의 생성자 외에는 (거의)없을 것 입니다. 결국 그럼 이 API는 잔소리 대상 두윗이 있는 유저 외에는 전부 cache miss 입니다.

gyun-ky replied:
> store라는게 read model의 성격을 지니는데, 조회 비용이 비싼 유스케이스의 성능을 높이기 위한 목적이 강하고 그 관점에서 속도가 빠른 저장소를 쓰는 것이기 때문에, store는 universal 하게 설계 할 부분은 아니라고 봅니다.

저는 여기서 Store에 대한 관점이 다르긴합니다
- 조회 비용이 비싼 유스케이스의 성능 높이는 목적은 Cache이고
- Redis를 저장소 목적에서 사용하는 것은 RDB는 지원 불가하고, Redis가 지원하는 자료구조(List, Map 등) + 비정형 데이터들에 지원을 받아서 원천 데이터 저장 or 비즈니스 플로우에서 중요한 요소 저장

으로 보고 있습니다

그런 차원에서 지금 계속 의견 일치가 안되는거 같긴하네요..!

gyun-ky replied:
> 또한 잔소리 대상 두윗이 캐싱된다고 하면, 두윗 단일 조회 API의 로직에 우선 redis 조회하도록 추가해야 하는데, 잔소리 대상 두윗을 상세조회 할 사람은 두윗의 생성자 외에는 (거의)없을 것 입니다. 결국 그럼 이 API는 잔소리 대상 두윗이 있는 유저 외에는 전부 cache miss 입니다.

유져 한명이 조회 한다고 cache 가치가 없는게 아니라고 생각합니다. 한명의 유져라도 비즈니스적으로 해당 데이터에 얼마나 많이 접근할지를 보는게 (hit rate)를 보는게 더 중요하지 않을까 싶습니다. 

더해서 잔소리 대상 Dowith 상세 조회 주체가 말씀하신것과 달리 많다고 분석됩니다
- Dowith 소유자
- Feed에 노출되어 N명의 유져

**Dowith 소유자 관점에서**
- 잔소리 대상 Dowith 인 경우 시작 시간이 지나서 N개의 잔소리 Push 알림이 온다
- 소유자는 지속적으로 push 알림을 눌러 Dowith 상세를 확인하게된다
- 소유자는 잔소리를 받기 싫어서 Dowith을 완료하게 되고 그때 상세를 또 확인하게 된다
=> 사실 상 이 Scene이 **Dowith을 게시하는 유져 입장에서는 제일 빠르고 편리해야한다고 생각합니다**

**Feed 유져 관점에서**
- N명이 Feed에서 지속적으로 잔소리 가능한 Dowith을 확인
- 클릭하여 상세를 확인하게 된다

S-J-Kim replied:
> 유져 한명이 조회 한다고 cache 가치가 없는게 아니라고 생각합니다. 한명의 유져라도 비즈니스적으로 해당 데이터에 얼마나 많이 접근할지를 보는게 (hit rate)를 보는게 더 중요하지 않을까 싶습니다.

hit rate는 서버 입장에서 API request 횟수 대비 cache hit인 응답이 얼마나 있는지로 계산을 해야지, 엔드 유저가 본인의 요청이 캐시를 얼마나 타는지로 판단할게 아니라고 봅니다. 그럼 내가 조회할 dowith이 캐시에 없는 유저한테는 miss rate가 100%인거잖아요.



> Dowith 소유자 관점에서
> 
> 잔소리 대상 Dowith 인 경우 시작 시간이 지나서 N개의 잔소리 Push 알림이 온다
> 소유자는 지속적으로 push 알림을 눌러 Dowith 상세를 확인하게된다
> 소유자는 잔소리를 받기 싫어서 Dowith을 완료하게 되고 그때 상세를 또 확인하게 된다
> => 사실 상 이 Scene이 Dowith을 게시하는 유져 입장에서는 제일 빠르고 편리해야한다고 생각합니다

저는 이 부분도 이해가 잘 안가는게, "두윗 상세"는 두윗의 정보 (기본 정보, 인증, 루틴) 을 조회하는 것이고 (잔소리 조회는 별도임) "실시간 잔소리 하기"는 (두윗 기본정보, 멤버정보, 잔소리 갯수) 를 조회합니다. 그래서 겹치는건 DowithTask 테이블 밖에 없는데, "universal 하게 구조를 저장해서 여러 api에서 같이 쓴다" 의 대상은 아니라고 봅니다.

두윗 목록 조회의 경우에도, 두윗 만 조회하는 경우는 없고, 투두와 같이 가져오는 API 밖에 없으니 무조건 DB를 거쳐야 합니다. 그러면 Redis를 타는 의미가 무색해지는거죠..

S-J-Kim replied:
캐싱하지 않는 것으로 결정하여 삭제 필요합니다. (CacheHelper테스트 수정 필요)

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/common/redis/CacheName.java:
```diff
@@ -1,11 +1,11 @@
-    public static final String DOWITH_TASK = "dowith";
-    public static final String LAZY_DOWITH_TASK = "lazy-dowith";
+    public static final String FEEDBACK_AVAILABLE_DOWITH_TASKS = "feed::dowithTask";
+    public static final String LAZY_DOWITH_TASK_IDS = "feed::dowithTask::lazy";
```
- LAZY_DOWITH_TASK_IDS -> LAZY_DOWITH_TASKS
- feed::dowithTask::lazy -> lazy-dowithTask

그리고 이건 Cache 가 아닌 StoreName으로 있어야하지 않을까 싶습니다

gyun-ky replied:
두개를 분리했을때 더 관리가 힘들려나요? 고민

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/domain/feed/model/FeedDowithTask.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * 둘러보기에서 두윗을 조회하기 위한 조회모델
+ */
+public record FeedDowithTask(
```
도메인 모델은 Class로 정의하는게 좋을거 같습니다

- Record는 레이어간 데이터 운반성으로 많이쓰이고 (Record의 탄생 배경도 참조했을 때) 
- 데이터 객체지향적으로 사용하기 위한 model은 class로 정의하는게 좋지 않을까요?

S-J-Kim replied:
네 저도 처음엔 클래스로 했다가 레코드로 바꾼것인데, Record는 equals 와 hashCode가 자동으로 생성되어 테스트에서 비교를 바로 할 수 있기 때문이었습니다.

그냥 클래스로 돌리고 이 두개를 오버라이드 해야겠네요

gyun-ky replied:
넵넵 그게 좋을거 같습니다~!

S-J-Kim replied:
캐시를 사용하지 않기록 결정하여 삭제되었습니다.

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/domain/feed/service/LazyDowithTaskSelector.java:
```diff
@@ -0,0 +1,18 @@
+@DomainService
+public class LazyDowithTaskSelector {
+
+    public List<FeedDowithTask> selectLazyDowithTasks(List<FeedDowithTask> dowithTasks, Integer size) {
```
몇개의 Lazy  Dowith을 select할 것인지는 정책이니까 size를 파라미터로 받는 것보다는 도메인 서비스 안에서 가지고 있는게 좋지 않을까요??

S-J-Kim replied:
넵,, 서비스 안에서 가지고 있는것이 맞을 것 같습니다.

gyun-ky replied:
조치되면 코멘트 남겨주세요

1. 리뷰대응 
2. 대응한 코멘트에 대응했다고 코멘트 달기
3. 리뷰어는 해당 사항 대응되었는지 확인
4. 리뷰어는 확인되면 resolve 처리

S-J-Kim replied:
수정 완료

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/feed/query/FeedQueryRepository.java:
```diff
@@ -0,0 +1,10 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public interface FeedQueryRepository {
```
여기도 DIP 원칙 적용 안된거 같습니다!

S-J-Kim replied:
옮기도록 하겠습니다 ㅋㅋ/...

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+     * @param key    식별자 (없을 경우 빈 문자열)
+     * @return prefix::key 형태의 full key
+     */
+    private String buildKey(RedisPolicySpec policy, String key) {
```
common > redis에 keyGenerator로 분리하면 좋을거 같습니다

gyun-ky replied:
그래서 CacheHelper랑 같이 사용할수 있도록요 

-> Key 생성 포맷의 통일화 by 단일 책임

S-J-Kim replied:
네.. key 자체는 String으로 반환하도록 해서 cache, store 둘다 사용 가능하도록 하겠습니다

S-J-Kim replied:
RedisKeyGenerator를 생성함

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+     * @param <T>       반환 타입
+     * @return Redis 값 또는 Optional.empty()
+     */
+    public <T> Optional<T> execute(Supplier<T> operation) {
```
만능 메서드 같은데 public이 맞나요?

S-J-Kim replied:
RedisOperator의 용도를 고려하면 이런 메서드는 private으로 내리는게 맞을 것 같습니다.

S-J-Kim replied:
수정 완료

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+     *
+     * @param operation Redis 수행 로직 (Lambda)
+     */
+    public void execute(Runnable operation) {
```
위 코멘트와 동일합니다

S-J-Kim replied:
수정 완료

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+     * @param fullKey Redis Full Key
+     * @param policy  CachePolicySpec
+     */
+    private void applyTtl(String fullKey, RedisPolicySpec policy) {
```
private 메서드들은 가독성을 위해 사용되는 메서드 아래나 파일의 아래쪽으로 위치되면 좋을거 같습니다!

S-J-Kim replied:
수정완료

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+        try {
+            operation.run();
+        } catch (DataAccessException e) {
+            log.warn("Redis Write/Delete Operation Failed. Cause: {}", e.getMessage());
```
exception 여기서 먹는거 같은데 먹으면 안되지 않을까요?

gyun-ky replied:
cache면 부가기능이니까 먹고 서버로그 ERROR로 남기는게 맞지만

Store로 Redis를 사용하는경우 이건 메인 비즈니스 로직 exception이기 때문에 위로 rasie되어야할거 같습니다

S-J-Kim replied:
Redis가 데이터의 원천이라면 예외가 서비스로 올라가야 하는게 맞지만, redis에서 (어떤 이유가 되었든) 가져온 데이터가 없다면 fallback을 수행하기 때문에 서비스로 올리지 않아도 될 것이라 판단했습니다. (Redis 장애에 애플리케이션이 영향받지 않아야 한다는 관점에서)

gyun-ky replied:
> Redis가 데이터의 원천이라면 예외가 서비스로 올라가야 하는게 맞지만,

Store 관점에서 RedisOperator를 사용하는거 아닌가요??

LAZY_DOWITH list 부터 보면, DB에는 없는 원천데이터로 저희가 정의하고 있는거 같은데요

> Redis 장애에 애플리케이션이 영향받지 않아야 한다는 관점

Redis에 들어있는게 원천데이터인 만큼 DB가 죽었을때 어플리케이션이 영향 받는것처럼 영향받을수 밖에 없고

fallback은 RedisOperator를 사용하는 service 단에서 처리해주는게 맞지 않을까 싶습니다 (ex. 어떤 서비스는 개발자한테 알람을 주고, 어떤 서비스는 DB에서 새롭게 데이터를 만들어서 서빙을 해준다거나)

S-J-Kim replied:
네, Operator 수준에서 예외는 전파하고 이걸 삼켜버릴지 아님 유저한테 오류를 뱉을지는 서비스 레이어에서 Operator의 사용 목적에 맞춰 판단하는 걸로 가시죠

S-J-Kim replied:
수정 완료

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+        execute(() -> {
+            if (redisTemplate.hasKey(fullOldKey)) {
+                redisTemplate.rename(fullOldKey, fullNewKey);
+                applyTtl(fullNewKey, policy);
```
위에 말씀드린대로 Store는 ttl을 policy에서 관리해서는 안되지 않을까 싶습니다

DB를 쓰듯이 우리가 서비스 로직에서 생명주기를 결정해야하는게 맞지 않나 싶습니다

- StorePolicy는 기본적으로 ttl X
- 서비스 비즈니스에 따라서 ttl을 쓰고 싶으면 RedisOperator 메서드 파라미터로 optional하게 받는게 맞지 않을까 싶습니다

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/redis/RedisOperator.java:
```diff
@@ -0,0 +1,347 @@
+    }
+
+    public <T> Optional<T> get(RedisPolicySpec policy, String key, Class<T> clazz) {
+        validatePolicy(policy, RedisValueType.STRING);
```
private 메서드는 this 붙이는게 가독성에 좀더 좋지 않을까 싶습니다 (이건 개인적인 생각입니다)

S-J-Kim replied:
네 저도 static method와 혼동할 수 있다고 봅니다.

S-J-Kim replied:
수정 완료

S-J-Kim committed:
feat: LazyDowithTaskSelector 에서 숫자는 받지 않도록 수정

S-J-Kim commented:
온라인 미팅 논의사항 정리
- Store 성격의 데이터는 Repository를 통해 접근하는 것으로 함 (구현 기술에 서비스가 영항받지 않으려는 의도)
- Feed API는 우선 캐싱하지 않는 것으로 결정함. (추후 필요시 API 전체 캐싱 -> Feedback count 분리 -> 배치를 통해 cache warmup 단계로 고도화) -> 따라서 배치는 삭제

S-J-Kim committed:
delete: 잔소리 대상 두윗 Redis 적재 배치 제거;

S-J-Kim committed:
delete: feed cache 관련 repository 제거; db 직접 쿼리로 변경

S-J-Kim committed:
delete: feed 도메인 모델 삭제

S-J-Kim committed:
refactor: repository interface를 application layer로 이관

S-J-Kim committed:
refactor: RedisOperator의 execute() 메서드 visibility 변경 to private

S-J-Kim committed:
feat: Redis key generator

S-J-Kim committed:
feat: RedisOperator execute() 시 Redis exception을 상위로 전파하도록 변경

S-J-Kim committed:
feat: 잔소리 대상 조회 로직에 페이징 추가, feedback count 조회 쿼리 분리

S-J-Kim committed:
feat: 잔소리 대상 두윗 조회 API

S-J-Kim committed:
feat: 잔소리 대상 두윗 조회 API 스펙 변경 및 디렉토리 이동;
- 내가 보낸 잔소리 목록 추가
- Feed -> Dowith 으로 이동

S-J-Kim committed:
test: 잔소리 대상 두윗 조회 API

S-J-Kim committed:
spotless

gyun-ky commented on src/main/java/com/LetMeDoWith/LetMeDoWith/application/task/dto/RetreiveFeedbackAvailableDowithTasksResult.java:
```diff
@@ -0,0 +1,51 @@
+import java.time.LocalTime;
+import java.util.List;
+
+public record RetreiveFeedbackAvailableDowithTasksResult(List<FeedbackAvailableDowithTaskDto> dowithTasks) {
```
Retrieve 오타요!

S-J-Kim replied:
수정 완료입니다 늦어서 죄송함다

S-J-Kim committed:
typo: retreive -> retrieve

gyun-ky commented:
제가 패키지 변경 건 엊그제에 머지해서 develop 이랑 conflict 해소 부탁드립니다

S-J-Kim committed:
refactor: 메서드 순서 변경 및 디렉토리 위치 변경

S-J-Kim committed:
Merge branch 'develop' into feature/TAS-434

# Conflicts:
#	src/main/java/com/LetMeDoWith/LetMeDoWith/application/task/service/RetrieveTaskService.java
#	src/main/java/com/LetMeDoWith/LetMeDoWith/domain/task/repository/DowithTaskQueryRepository.java
#	src/main/java/com/LetMeDoWith/LetMeDoWith/domain/task/repository/dto/FeedbackAvailableDowithTasksQueryDto.java
#	src/main/java/com/LetMeDoWith/LetMeDoWith/infrastructure/task/query/DowithTaskQueryRespositoryImpl.java
#	src/main/java/com/LetMeDoWith/LetMeDoWith/presentation/task/controller/DowithTaskController.java

S-J-Kim committed:
refactor: 메서드 순서 변경 및 디렉토리 위치 변경
